### ISSUE TITLE
Error in similarity score?

### ISSUE THREAD

**Issue by dwolfeu** [benfred/implicit#205]
I think there is an error in the calculation of the similarity score. I first noticed the problem because I was getting scores larger than 1 and even a score different from 1 for the similarity of an item with itself. I believe I have found the source of the error.

In `recommender_base.py` the similarity score is defined by the following function:

    def _get_similarity_score(self, factor, factors, norms, N):
        scores = factors.dot(factor) / norms
        best = np.argpartition(scores, -N)[-N:]
        return sorted(zip(best, scores[best]), key=lambda x: -x[1]

The error here is that the dot product is being divided by only _one_ of the norms. I think the following definition is correct:

    def _get_similarity_score(self, factor, factors, norms, N):
        scores = factors.dot(factor) / (np.linalg.norm(factor)*norms)
        best = np.argpartition(scores, -N)[-N:]

        return sorted(zip(best, scores[best]), key=lambda x: -x[1]

I have tested it locally and it works. I would have already sent a pull request, but I don't have permission to push to the repository.

**Comment by dwolfeu** [comment]
In fact, the following should be a more efficient correction:

    def _get_similarity_score(self, factor, factors, norms, N):
        scores = factors.dot(factor) / norms
        best = np.argpartition(scores, -N)[-N:]
    
        return sorted(zip(best, scores[best]/np.linalg.norm(factor)), key=lambda x: -x[1])

This is how it is in version 0.3.4.

**Comment by ita9naiwa** [comment]
It will not sum up to one, but relative order is same if we didn't multiply the norm of `factor`.

I don't know exactly why norm of `factor` is not there, but probably because it requires additional operation of cost of O(N), but results are same.

I think

```
    scores = factors.dot(factor) / norms
    best = np.argpartition(scores, -N)[-N:]
    scores = scores / np.sqrt(np.sum(factor, factor))
```
this will be somewhat faster, btw.

**Comment by dwolfeu** [comment]
I would be surprised if it was removed because of increased cost, since it's only O(N). Moreover, the similarity score was calculated correctly in version 0.3.4. (I think the error was introduced in the change from 0.3.6 to 0.3.7.)

Although the relative rankings will be the same for any given item, relative cosine similarities are easier to interpret. More importantly, the currently incorrect calculation can lead to error. A toy example:

item A: (5, 10)
item B: (2, 1.1)
item C: (2, 1)

**Incorrect calculation of similarity score**

similarity_score(A, C) = (5 * 2 + 10 * 1) / norm(C) = 8.9 (approx.)
similarity_score(B, C) = (2 * 2 + 1.1 * 1) / norm(C) = 2.3 (approx.)

False conclusion: Item C is more similar to item A than it is to item B.

**Correct calculation of similarity score**

similarity_score(A, C) = (5 * 2 + 10 * 1) / (norm(A)*norm(C)) =  0.8
similarity_score(B, C) = (2 * 2 + 1.1 * 1) / (norm(B)*norm(C)) = 1.0 (approx.)

Correct conclusion: Item C is more similar to item B than it is to item A.

So, in short, I really think is this error should be fixed. I am happy to send a pull request (although I would to be given permission to push to the repository).

**Comment by Acey25** [comment]
Thank goodness @dwolfeu that you had opened this issue!  I wish I had checked the issues before spending a couple hours trying to understand why my scores were more than 1! I hope @benfred pushes this fix to the repo soon.

**Comment by ita9naiwa** [comment]
Your calculation is wrong
incorrect
```
similarity_score(A, C) = (5 * 2 + 10 * 1)
similarity_score(B, C) = (2 * 2 + 1.1 * 1)
```

correct
```
similarity_score(A, C) = (5 * 2 + 10 * 1) / norm(A)
similarity_score(B, C) = (2 * 2 + 1.1 * 1) / norm(B)
```

Relative order is conserved
variable `norms` is list of norms.

**Comment by dwolfeu** [comment]
With the cosine similarity one divides by the product of _both_ of the norms: https://en.wikipedia.org/wiki/Cosine_similarity

**Comment by ita9naiwa** [comment]
I know but norm of one element is same for the all ranking calcs so resultant rank does not change...

**Comment by dwolfeu** [comment]
That's okay if one calculates similarity scores for only one item, but what if one calculates similarity scores for one item and then again for another item? This was the idea with the toy example above.

Also, it's confusing for people if they think they are calculating the cosine similarity, only to get a similarity score higher than 1. See the reply from @AsiehH above, for example.

Anyway, I've corrected the error on my local installation. I just think others would benefit from the source code being fixed too.

**Comment by ita9naiwa** [comment]
I got it. I think your point makes sense.
There'll be some situations that things happen.

**Comment by aditya-malte** [comment]
Hello,
I'd suggest dividing by the norm of the current element even if the order remains the same, for explainability purpose. (spent around an hour figuring out how cosine similarity was used.) Plus, the scores appear too small(0.09,0.065,etc.) without dividing by magnitude of the comparing course. This might be an issue if you are using(and comparing) the scores somehow.

Thanks

**Comment by bruce0715** [comment]
I just wonder if this problem has already fixed? I use pip install to install implicit. When I calculate the similarity between items, the item itself still doesn't equal to 1.

**Comment by aditya-malte** [comment]
I think I should make a pull request to solve this issue. It’s really confusing for the end-user(because of this issue) of the library to guess what is going on in the backend to compute similarities.

**Comment by bruce0715** [comment]
> I think I should make a pull request to solve this issue. It’s really confusing for the end-user(because of this issue) of the library to guess what is going on in the backend to compute similarities.

That will be great! I want to change locally by myself, but I have no idea how to deal with the .so file.

**Comment by SavanGowda** [comment]
@bruce0715 I think that you could fork the repo, and do the necessary changes locally and you can pip install the repo 

for example - **pip install git+_add the repo url here_**

I hope that helps :)
